keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,522 @@
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposal introduces two new sig-api-machinery subprojects to improve the experience of developing Kubernetes APIs (including CRDs).
+
+A linter, `kube-api-linter` (aka `kal`) as a golangci-lint plugin for evaluating Go types that are used to generate both built-in types and CRDs, to ensure they follow the Kubernetes API conventions and best practices.
+
+A CRD schema upgrade checker, `crdify` as a CLI that compares generated CRD schemas to identify changes to the CRD schema, and ensure that any changes are compatible.
+
+## Motivation
+
+Aside from existing documentation on [Kubernetes API conventions][kube-api-conventions] and [making changes to Kubernetes APIs][making-changes-to-apis],
+there is little to no tooling that enable developers to ensure they are following best practices when developing Kubernetes-native APIs.
+
+This KEP aims to improve the Kubernetes-native API development experience by adding two sig-api-machinery subprojects for:
+
+- Linting Kubernetes APIs written in Go based on the [Kubernetes API conventions][kube-api-conventions]
+- Validating changes to CustomResourceDefinition schemas in YAML
+
+[kube-api-conventions]:(https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+[making-changes-to-apis]:(https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+
+### Goals
+
+- Create tooling to help developers writing Kubernetes APIs in Go to follow best practices and Kubernetes API conventions
```

Comment:
This is excellent.

Having a "delegate" from @kubernetes/api-reviewers might valuable here. (Any volunteers to do a pass and provide guidance feedback @liggitt @msau42 @deads2k @thockin)?

cc @yongruilin, @aaron-prindle who has been working on in-tree validation linting for Declarative Validation

Reply:
I'd be keen to catch up with @yongruilin and @aaron-prindle about their progress. So far I've been focussing the linters as much as possible on those that don't need markers for validation, or where kube markers are also supported for CRDs.

There are some examples now though (`maxlength`) where it checks for a maximal string/list length, that only now supports CRDs. I'd hope in the future to be able to extend that linter to be able to cover built-in types too. Understanding the goals of the in-tree validation would help me understand how to drive KAL towards being useful for both built-ins and CRDs

Reply:
+1 on the delegate from the kubernetes/api-reviewers group. I've got some text in the "Risks and Mitigation" section below that calls out the need for a robust review process with "stakeholders". I'll add an entry to that list.